### PR TITLE
Add diagnostic logs for authentication flows

### PR DESCRIPTION
**Pull request by chamathns**
### Proposed changes in this pull request

[List all changes you want to add here. If you fixed an issue, please
add a reference to that issue as well.]

-

### When should this PR be merged

[Please describe any preconditions that need to be addressed before we
can merge this pull request.]

### Follow up actions

[List any possible follow-up actions here; for instance, testing data
migrations, software that we need to install on staging and production
environments.]

-

#### General

- [ ] **Is this PR explained thoroughly?** All code changes must be accounted for in the PR description.
- [ ] **Is the PR labeled correctly?**

#### Functionality

- [ ] **Are all requirements met?** Compare implemented functionality with the requirements specification.
- [ ] **Does the UI work as expected?** There should be no Javascript errors in the console; all resources should load. There should be no unexpected errors. Deliberately try to break the feature to find out if there are corner cases that are not handled.

#### Code

- [ ] **Do you fully understand the introduced changes to the code?** If not ask for clarification, it might uncover ways to solve a problem in a more elegant and efficient way.
- [ ] **Does the PR introduce any inefficient database requests?** Use the debug server to check for duplicate requests.
- [ ] **Are all necessary strings marked for translation?** All strings that are exposed to users via the UI must be [marked for translation](https://docs.djangoproject.com/en/1.10/topics/i18n/translation/).

#### Tests

- [ ] **Are there sufficient test cases?** Ensure that all components are tested individually; models, forms, and serializers should be tested in isolation even if a test for a view covers these components.
- [ ] **If this is a bug fix, are tests for the issue in place?**  There must be a test case for the bug to ensure the issue won’t regress. Make sure that the tests break without the new code to fix the issue.
- [ ] **If this is a new feature or a significant change to an existing feature?** has the manual testing spreadsheet been updated with instructions for manual testing?

#### Security

- [ ] **Confirm this PR doesn't commit any keys, passwords, tokens, usernames, or other secrets.**
- [ ] **Are all UI and API inputs run through forms or serializers?**
- [ ] **Are all external inputs validated and sanitized appropriately?**
- [ ] **Does all branching logic have a default case?**
- [ ] **Does this solution handle outliers and edge cases gracefully?**
- [ ] **Are all external communications secured and restricted to SSL?**

#### Documentation

- [ ] **Are changes to the UI documented in the platform docs?** If this PR introduces new platform site functionality or changes existing ones, the changes should be documented.
- [ ] **Are changes to the API documented in the API docs?** If this PR introduces new API functionality or changes existing ones, the changes must be documented.
- [ ] **Are reusable components documented?** If this PR introduces components that are relevant to other developers (for instance a mixin for a view or a generic form) they should be documented in the Wiki.

### CODE REVIEW

File: components/authentication-framework/org.wso2.carbon.identity.application.authentication.framework/src/main/java/org/wso2/carbon/identity/application/authentication/framework/handler/request/impl/DefaultRequestCoordinator.java [hunk 1]
```diff
@@ -167,6 +170,13 @@ public void handle(HttpServletRequest request, HttpServletResponse response) thr
                             log.debug("No authentication request found in the cache for sessionDataKey: "
                                     + sessionDataKey);
                         }
+                        if (LoggerUtils.isDiagnosticLogsEnabled()) {
+                            LoggerUtils.triggerDiagnosticLogEvent(
+                                    FrameworkConstants.LogConstants.AUTHENTICATION_FRAMEWORK, null,
+                                    LogConstants.SUCCESS, "No authentication request found in the " +
+                                            "cache for sessionDataKey: " + sessionDataKey, "handle-authentication" +
```
Comment (pulasthi7) [inline components/authentication-framework/org.wso2.carbon.identity.application.authentication.framework/src/main/java/org/wso2/carbon/identity/application/authentication/framework/handler/request/impl/DefaultRequestCoordinator.java:177]: feels this is more technical, cache is an internal thing. so, better to reword this to convey what happened from external user pov

File: components/authentication-framework/org.wso2.carbon.identity.application.authentication.framework/src/main/java/org/wso2/carbon/identity/application/authentication/framework/handler/request/impl/DefaultRequestCoordinator.java [hunk 2]
```diff
@@ -185,6 +195,13 @@ public void handle(HttpServletRequest request, HttpServletResponse response) thr
                     if (log.isDebugEnabled()) {
                         log.debug("Session data key is null in the request and not a logout request.");
                     }
+                    if (LoggerUtils.isDiagnosticLogsEnabled()) {
+                        LoggerUtils.triggerDiagnosticLogEvent(
+                                FrameworkConstants.LogConstants.AUTHENTICATION_FRAMEWORK, null, LogConstants.FAILED,
+                                "Not a logout request and session data key " +
+                                        "is null in the request. Sending to retry page.", "handle-authentication" +
```
Comment (pulasthi7) [inline components/authentication-framework/org.wso2.carbon.identity.application.authentication.framework/src/main/java/org/wso2/carbon/identity/application/authentication/framework/handler/request/impl/DefaultRequestCoordinator.java:202]: better to reword a bit here as well. also lets have the exact param name "sessionDataKey" so that they can debug easily. And better to say sdk is "not provided or is null in the request"

File: components/authentication-framework/org.wso2.carbon.identity.application.authentication.framework/src/main/java/org/wso2/carbon/identity/application/authentication/framework/handler/request/impl/DefaultRequestCoordinator.java [hunk 3]
```diff
@@ -227,6 +244,18 @@ public void handle(HttpServletRequest request, HttpServletResponse response) thr
                                     "Request Headers: " + getHeaderString(request) + "\n" +
                                     "Thread Id: " + Thread.currentThread().getId());
                         }
+                        if (LoggerUtils.isDiagnosticLogsEnabled()) {
+                            Map<String, Object> params = new HashMap<>();
+                            params.put("Context id", context.getContextIdentifier());
+                            params.put("Originating address", request.getRemoteAddr());
+                            params.put("Request Headers", getHeaderString(request));
+                            params.put("Thread Id", Thread.currentThread().getId());
```
Comment (pulasthi7) [inline components/authentication-framework/org.wso2.carbon.identity.application.authentication.framework/src/main/java/org/wso2/carbon/identity/application/authentication/framework/handler/request/impl/DefaultRequestCoordinator.java:252]: let's have constants for these keys and reuse as much as possible

File: components/authentication-framework/org.wso2.carbon.identity.application.authentication.framework/src/main/java/org/wso2/carbon/identity/application/authentication/framework/handler/request/impl/DefaultRequestCoordinator.java [hunk 4]
```diff
@@ -227,6 +244,18 @@ public void handle(HttpServletRequest request, HttpServletResponse response) thr
...
+                            LoggerUtils.triggerDiagnosticLogEvent(
+                                    FrameworkConstants.LogConstants.AUTHENTICATION_FRAMEWORK, params,
+                                    LogConstants.FAILED, "Same context is currently in used by a " +
```
Comment (pulasthi7) [inline components/authentication-framework/org.wso2.carbon.identity.application.authentication.framework/src/main/java/org/wso2/carbon/identity/application/authentication/framework/handler/request/impl/DefaultRequestCoordinator.java:255]: again better to reword to something like "same authentication request is being processed by another thread. Could be a possible double submit or replay request"

File: components/authentication-framework/org.wso2.carbon.identity.application.authentication.framework/src/main/java/org/wso2/carbon/identity/application/authentication/framework/handler/request/impl/DefaultRequestCoordinator.java (hunk 2 above)
Comment (chamathns) [reply components/authentication-framework/org.wso2.carbon.identity.application.authentication.framework/src/main/java/org/wso2/carbon/identity/application/authentication/framework/handler/request/impl/DefaultRequestCoordinator.java:202]: *session data key parameter (add the request parameter as it is) is not provided in the request.

Comment (chamathns) [reply components/authentication-framework/org.wso2.carbon.identity.application.authentication.framework/src/main/java/org/wso2/carbon/identity/application/authentication/framework/handler/request/impl/DefaultRequestCoordinator.java]: Addressed with https://github.com/wso2/carbon-identity-framework/pull/4326/commits/ca34cc4be2ac20f9c870f1ded78fab17d9bf6d94

File: components/authentication-framework/org.wso2.carbon.identity.application.authentication.framework/src/main/java/org/wso2/carbon/identity/application/authentication/framework/handler/request/impl/DefaultRequestCoordinator.java (hunk 1 above)
Comment (chamathns) [reply components/authentication-framework/org.wso2.carbon.identity.application.authentication.framework/src/main/java/org/wso2/carbon/identity/application/authentication/framework/handler/request/impl/DefaultRequestCoordinator.java]: This has been removed with https://github.com/wso2/carbon-identity-framework/pull/4326/commits/ca34cc4be2ac20f9c870f1ded78fab17d9bf6d94 since this information is internal and not relevant for the user.

File: components/authentication-framework/org.wso2.carbon.identity.application.authentication.framework/src/main/java/org/wso2/carbon/identity/application/authentication/framework/handler/request/impl/DefaultRequestCoordinator.java (hunk 4 above)
Comment (chamathns) [reply components/authentication-framework/org.wso2.carbon.identity.application.authentication.framework/src/main/java/org/wso2/carbon/identity/application/authentication/framework/handler/request/impl/DefaultRequestCoordinator.java]: Addressed with https://github.com/wso2/carbon-identity-framework/pull/4326/commits/c15402ecaefef87740033c886e79165563e32b8f

File: components/authentication-framework/org.wso2.carbon.identity.application.authentication.framework/src/main/java/org/wso2/carbon/identity/application/authentication/framework/handler/request/impl/PostAuthnMissingClaimHandler.java [hunk 5]
```diff
@@ -197,6 +209,17 @@ protected PostAuthnHandlerFlowStatus handlePostAuthenticationForMissingClaimsReq
...
                     Claim claimObj = claimManager.getClaim(missingClaim.getValue());
                     if (claimObj != null && claimObj.isReadOnly()) {
+                        if (LoggerUtils.isDiagnosticLogsEnabled()) {
+                            Map<String, Object> params = new HashMap<>();
+                            params.put(FrameworkConstants.LogConstants.SERVICE_PROVIDER,
+                                    context.getServiceProviderName());
+                            params.put(FrameworkConstants.LogConstants.TENANT_DOMAIN, context.getTenantDomain());
+                            LoggerUtils.triggerDiagnosticLogEvent(
+                                    FrameworkConstants.LogConstants.AUTHENTICATION_FRAMEWORK, params,
+                                    LogConstants.FAILED, "One or more read-only claim is missing in the requested " +
```
Comment (pulasthi7) [inline components/authentication-framework/org.wso2.carbon.identity.application.authentication.framework/src/main/java/org/wso2/carbon/identity/application/authentication/framework/handler/request/impl/PostAuthnMissingClaimHandler.java:219]: can we say which are missing?

File: components/authentication-framework/org.wso2.carbon.identity.application.authentication.framework/src/main/java/org/wso2/carbon/identity/application/authentication/framework/handler/request/impl/PostAuthnMissingClaimHandler.java [hunk 6]
```diff
@@ -288,6 +311,15 @@ protected void handlePostAuthenticationForMissingClaimsResponse(HttpServletReque
 
         if (!doMandatoryClaimsExist) {
             // Check whether mandatory claims exist in the request. If not throw error.
+            if (LoggerUtils.isDiagnosticLogsEnabled()) {
+                Map<String, Object> params = new HashMap<>();
+                params.put(FrameworkConstants.LogConstants.SERVICE_PROVIDER, context.getServiceProviderName());
+                params.put(FrameworkConstants.LogConstants.TENANT_DOMAIN, context.getTenantDomain());
+                LoggerUtils.triggerDiagnosticLogEvent(
+                        FrameworkConstants.LogConstants.AUTHENTICATION_FRAMEWORK, params, LogConstants.FAILED,
+                        "Mandatory missing claims are not found",
```
Comment (pulasthi7) [inline components/authentication-framework/org.wso2.carbon.identity.application.authentication.framework/src/main/java/org/wso2/carbon/identity/application/authentication/framework/handler/request/impl/PostAuthnMissingClaimHandler.java:320]: Same here.. can we say which are missing?

File: components/authentication-framework/org.wso2.carbon.identity.application.authentication.framework/src/main/java/org/wso2/carbon/identity/application/authentication/framework/handler/request/impl/consent/ConsentMgtPostAuthnHandler.java [hunk 7]
```diff
@@ -185,14 +187,28 @@ protected PostAuthnHandlerFlowStatus handlePreConsent(HttpServletRequest request
...
+                    params.put(FrameworkConstants.LogConstants.MANDATORY_CLAIMS,
+                            consentClaimsData.getMandatoryClaims());
+                    params.put(FrameworkConstants.LogConstants.REQUESTED_CLAIMS,
+                            consentClaimsData.getRequestedClaims());
+                    params.put(FrameworkConstants.LogConstants.USER,
+                            authenticatedUser.getAuthenticatedSubjectIdentifier());
+                    params.put(FrameworkConstants.LogConstants.SERVICE_PROVIDER, serviceProvider.getApplicationName());
+                    params.put(FrameworkConstants.LogConstants.TENANT_DOMAIN, getSPTenantDomain(serviceProvider));
+                    LoggerUtils.triggerDiagnosticLogEvent(FrameworkConstants.LogConstants.AUTHENTICATION_FRAMEWORK,
+                            params, LogConstants.SUCCESS, "Require consent for mandatory claims ",
```
Comment (pulasthi7) [inline components/authentication-framework/org.wso2.carbon.identity.application.authentication.framework/src/main/java/org/wso2/carbon/identity/application/authentication/framework/handler/request/impl/consent/ConsentMgtPostAuthnHandler.java:209]: Is this only for mandatory claims? or all claims

File: components/authentication-framework/org.wso2.carbon.identity.application.authentication.framework/src/main/java/org/wso2/carbon/identity/application/authentication/framework/handler/request/impl/PostAuthnMissingClaimHandler.java (hunk 5 above)
Comment (chamathns) [reply components/authentication-framework/org.wso2.carbon.identity.application.authentication.framework/src/main/java/org/wso2/carbon/identity/application/authentication/framework/handler/request/impl/PostAuthnMissingClaimHandler.java:219]: this loops through the missing claims and throws an error when the first read-only claim is found. We can log that claim here

File: components/authentication-framework/org.wso2.carbon.identity.application.authentication.framework/src/main/java/org/wso2/carbon/identity/application/authentication/framework/handler/request/impl/PostAuthnMissingClaimHandler.java (hunk 6 above)
Comment (chamathns) [reply components/authentication-framework/org.wso2.carbon.identity.application.authentication.framework/src/main/java/org/wso2/carbon/identity/application/authentication/framework/handler/request/impl/PostAuthnMissingClaimHandler.java:320]: In this case, we are checking the response params to see if we have the mandatory params included there. These are sent when the user is prompted to enter the mandatory params. So, the handler knows it must return at least one mandatory claim in the response. The way it sends the mandatory claims is by prepending the prefix : `claim_mand~`
If the response does not contain any params, the handler will throw an error and break the flow.
Therefore, the missing params are nort available in the context here

File: components/authentication-framework/org.wso2.carbon.identity.application.authentication.framework/src/main/java/org/wso2/carbon/identity/application/authentication/framework/handler/request/impl/consent/ConsentMgtPostAuthnHandler.java (hunk 7 above)
Comment (chamathns) [reply components/authentication-framework/org.wso2.carbon.identity.application.authentication.framework/src/main/java/org/wso2/carbon/identity/application/authentication/framework/handler/request/impl/consent/ConsentMgtPostAuthnHandler.java:209]:
```
{"logId":"8b262f48-f298-4192-8d73-4605bb6a4c5e","recordedAt":{"seconds":1668778734,"nanos":807590000},"requestId":"7248fd35-072f-49c0-b64c-cf2ce1e52776","resultStatus":"SUCCESS","resultMessage":"Require consent for mandatory claims ","actionId":"process-claim-consent","componentId":"authentication-framework","input":{"tenant domain":"risekite.com","service provider":"sample-app","requested claims":[{"id":0,"claimUri":"http://wso2.org/claim/country","displayName":"Country","description":"Country"}],"user":"aca7acd9-5b7d-49d1-9757-308352aa2072","mandatory claims":[]}}
```

File: components/authentication-framework/org.wso2.carbon.identity.application.authentication.framework/src/main/java/org/wso2/carbon/identity/application/authentication/framework/handler/request/impl/PostAuthnMissingClaimHandler.java (hunk 5 above)
Comment (chamathns) [reply components/authentication-framework/org.wso2.carbon.identity.application.authentication.framework/src/main/java/org/wso2/carbon/identity/application/authentication/framework/handler/request/impl/PostAuthnMissingClaimHandler.java:220]: Addressed with https://github.com/wso2/carbon-identity-framework/pull/4326/commits/4244d1316c95005cd9d8181ba51c04f371175f2e

File: components/authentication-framework/org.wso2.carbon.identity.application.authentication.framework/src/main/java/org/wso2/carbon/identity/application/authentication/framework/handler/request/impl/consent/ConsentMgtPostAuthnHandler.java (hunk 7 above)
Comment (pulasthi7) [reply components/authentication-framework/org.wso2.carbon.identity.application.authentication.framework/src/main/java/org/wso2/carbon/identity/application/authentication/framework/handler/request/impl/consent/ConsentMgtPostAuthnHandler.java:209]: Then can we remove the mandatory part in the log message?

File: components/authentication-framework/org.wso2.carbon.identity.application.authentication.framework/src/main/java/org/wso2/carbon/identity/application/authentication/framework/handler/step/impl/DefaultStepHandler.java [hunk 8]
```diff
@@ -263,6 +280,16 @@ public void handle(HttpServletRequest request, HttpServletResponse response,
...
                         LOG.debug("Already authenticated. Skipping the step");
                     }
+                    if (LoggerUtils.isDiagnosticLogsEnabled()) {
+                        Map<String, Object> params = new HashMap<>();
+                        params.put(FrameworkConstants.LogConstants.SERVICE_PROVIDER, context.getServiceProviderName());
+                        params.put(FrameworkConstants.LogConstants.TENANT_DOMAIN, context.getTenantDomain());
+                        params.put(FrameworkConstants.LogConstants.STEP, stepConfig.getOrder());
+                        LoggerUtils.triggerDiagnosticLogEvent(
+                                FrameworkConstants.LogConstants.AUTHENTICATION_FRAMEWORK, params, LogConstants.SUCCESS,
+                                "Already authenticated. Skipping the step",
```
Comment (pulasthi7) [inline components/authentication-framework/org.wso2.carbon.identity.application.authentication.framework/src/main/java/org/wso2/carbon/identity/application/authentication/framework/handler/step/impl/DefaultStepHandler.java:290]: can we say with which IDP/ authenticator the authentication is already complete?

File: components/authentication-framework/org.wso2.carbon.identity.application.authentication.framework/src/main/java/org/wso2/carbon/identity/application/authentication/framework/handler/step/impl/DefaultStepHandler.java [hunk 9]
```diff
@@ -710,10 +748,33 @@ protected void doAuthentication(HttpServletRequest request, HttpServletResponse
...
             }
+            if (LoggerUtils.isDiagnosticLogsEnabled()) {
+                 Map<String, Object> params = getContextParamsForDiagnosticLogs(context, authenticatorConfig,
+                        stepConfig);
+                Optional.of(e.getUser()).ifPresent(user -> {
+                    params.put(FrameworkConstants.LogConstants.USER, user.getLoggableUserId());
+                    params.put(FrameworkConstants.LogConstants.USER_STORE_DOMAIN, user.getUserStoreDomain());
+                });
+                LoggerUtils.triggerDiagnosticLogEvent(FrameworkConstants.LogConstants.AUTHENTICATION_FRAMEWORK, params,
+                        LogConstants.FAILED, "Authentication failed: " + e.getMessage(),
```
Comment (pulasthi7) [inline components/authentication-framework/org.wso2.carbon.identity.application.authentication.framework/src/main/java/org/wso2/carbon/identity/application/authentication/framework/handler/step/impl/DefaultStepHandler.java:775]: can we say with which authenticator/idp?

File: components/authentication-framework/org.wso2.carbon.identity.application.authentication.framework/src/main/java/org/wso2/carbon/identity/application/authentication/framework/handler/request/impl/consent/ConsentMgtPostAuthnHandler.java (hunk 7 above)
Comment (chamathns) [reply components/authentication-framework/org.wso2.carbon.identity.application.authentication.framework/src/main/java/org/wso2/carbon/identity/application/authentication/framework/handler/request/impl/consent/ConsentMgtPostAuthnHandler.java]: +1. removed the "mandatory" part with https://github.com/wso2/carbon-identity-framework/pull/4326/commits/9a1423aa5dca4a529cd5a0abe0c755ac81eee53c

File: components/authentication-framework/org.wso2.carbon.identity.application.authentication.framework/src/main/java/org/wso2/carbon/identity/application/authentication/framework/handler/step/impl/DefaultStepHandler.java (hunk 8 above)
Comment (chamathns) [reply components/authentication-framework/org.wso2.carbon.identity.application.authentication.framework/src/main/java/org/wso2/carbon/identity/application/authentication/framework/handler/step/impl/DefaultStepHandler.java:297]: Added with https://github.com/wso2/carbon-identity-framework/pull/4326/commits/9a1423aa5dca4a529cd5a0abe0c755ac81eee53c

File: components/authentication-framework/org.wso2.carbon.identity.application.authentication.framework/src/main/java/org/wso2/carbon/identity/application/authentication/framework/handler/step/impl/DefaultStepHandler.java (hunk 9 above)
Comment (chamathns) [reply components/authentication-framework/org.wso2.carbon.identity.application.authentication.framework/src/main/java/org/wso2/carbon/identity/application/authentication/framework/handler/step/impl/DefaultStepHandler.java:785]: +1. Added with https://github.com/wso2/carbon-identity-framework/pull/4326/commits/9a1423aa5dca4a529cd5a0abe0c755ac81eee53c

**Review by pulasthi7 — APPROVED** [review]

**Review by pulasthi7 — APPROVED** [review]
